### PR TITLE
Implement error_handler function in libsyclinterface/helper 

### PR DESCRIPTION
**Pull request by vlad-perevezentsev**
This PR adds two overloaded `error_handler` functions to pass an exception object and output an error.
Change all functions in `dpctl_sycl_plaftorm_interface.cpp` to use the `error_handler`.

### CODE REVIEW

File: libsyclinterface/source/dpctl_sycl_platform_interface.cpp [hunk 1]
```diff
@@ -50,15 +51,16 @@ DPCTLPlatform_Copy(__dpctl_keep const DPCTLSyclPlatformRef PRef)
...
+        error_handler("Cannot copy DPCTLSyclPlatformRef as input is a nullptr.",
+                      __FILE__, __func__, __LINE__);
         return nullptr;
     }
     try {
         auto CopiedPlatform = new platform(*Platform);
         return wrap(CopiedPlatform);
-    } catch (std::bad_alloc const &ba) {
+    } catch (std::exception const &e) {
         // \todo log error
```
Comment (diptorupd) [inline libsyclinterface/source/dpctl_sycl_platform_interface.cpp:62]: Please delete all the `/todo` tags as that is finally getting implemented 😄

File: libsyclinterface/source/dpctl_sycl_platform_interface.cpp [hunk 2]
```diff
@@ -84,120 +87,125 @@ __dpctl_give DPCTLSyclPlatformRef DPCTLPlatform_CreateFromSelector(
...
         }
-    }
-    else {
-        std::cerr << "Device selector pointer cannot be NULL\n";
-    }
-
-    return nullptr;
-}
+        else {
+            error_handler("Device selector pointer cannot be.", __FILE__,
```
Comment (oleksandr-pavlyk) [inline libsyclinterface/source/dpctl_sycl_platform_interface.cpp]: cannot be NULL.

File: libsyclinterface/source/dpctl_sycl_platform_interface.cpp [hunk 3]
```diff
@@ -84,120 +87,125 @@ __dpctl_give DPCTLSyclPlatformRef DPCTLPlatform_CreateFromSelector(
         try {
             P = new platform(*DS);
             return wrap(P);
-        } catch (std::bad_alloc const &ba) {
-            std::cerr << ba.what() << '\n';
-            return nullptr;
         } catch (runtime_error const &re) {
             delete P;
-            std::cerr << re.what() << '\n';
+            error_handler(re, __FILE__, __func__, __LINE__);
```
Comment (oleksandr-pavlyk) [inline libsyclinterface/source/dpctl_sycl_platform_interface.cpp:91]: ``sycl::runtime_error`` is deprecated, and is also an instance of class derived from `std::exception``

File: libsyclinterface/helper/source/dpctl_error_handlers.cpp [hunk 4]
```diff
@@ -40,3 +41,31 @@ void DPCTL_AsyncErrorHandler::operator()(
...
+
+void error_handler(const std::exception &e,
+                   const char *file_name,
+                   const char *func_name,
+                   int line_num,
+                   int level)
+{
+    const char *verbose = std::getenv("DPCTL_VERBOSITY");
+    if (verbose) {
+        if (!strncmp(verbose, "error", 5))
```
Comment (oleksandr-pavlyk) [inline libsyclinterface/helper/source/dpctl_error_handlers.cpp:53]: I think we should define `int requested_level = 0;`, then change it depending on the env variable. 

At the end we should do

```c++
   if (requested_level <= level) {
        std::cerr << e.what() << " in " << file_name << " at " << func_name
                        << " on line: " << line_num << std::endl;
   }
```

File: libsyclinterface/source/dpctl_sycl_context_interface.cpp [hunk 5]
```diff
@@ -43,28 +43,25 @@ DEFINE_SIMPLE_CONVERSION_FUNCTIONS(std::vector<DPCTLSyclDeviceRef>,
 
 __dpctl_give DPCTLSyclContextRef
 DPCTLContext_Create(__dpctl_keep const DPCTLSyclDeviceRef DRef,
-                    error_handler_callback *error_handler,
+                    error_handler_callback *handler,
                     int /**/)
 {
     DPCTLSyclContextRef CRef = nullptr;
     auto Device = unwrap(DRef);
     if (!Device)
```
Comment (oleksandr-pavlyk) [inline libsyclinterface/source/dpctl_sycl_context_interface.cpp:51]: there should be an `error_handler` call here.

File: libsyclinterface/source/dpctl_sycl_device_manager.cpp [hunk 6]
```diff
@@ -134,8 +135,9 @@ struct DeviceCacheBuilder
                     auto entry = cache_l.emplace(D, Ctx);
 
                     if (!entry.second) {
-                        std::cerr << "Fatal Error during device cache "
-                                     "construction.\n";
+                        error_handler("Fatal Error during device cache "
+                                      "construction.",
+                                      __FILE__, __func__, __LINE__);
```
Comment (oleksandr-pavlyk) [inline libsyclinterface/source/dpctl_sycl_device_manager.cpp:140]:
```suggestion
                                      __FILE__, __func__, __LINE__, /* level=always */ 0);
```

File: libsyclinterface/helper/source/dpctl_error_handlers.cpp [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+            requested_type_level = error_level::none;
+        else if (!strncmp(verbose, "error", 5))
+            requested_type_level = error_level::error;
+        else if (!strncmp(verbose, "warning", 7))
+            requested_type_level = error_level::warning;
+    }
+
+    if (requested_type_level <= type_error) {
+        std::cerr << e.what() << " in " << func_name << " at " << file_name
+                  << ": " << line_num << std::endl;
```
Comment (oleksandr-pavlyk) [inline libsyclinterface/helper/source/dpctl_error_handlers.cpp:65]:
```suggestion
                  << ":" << line_num << std::endl;
```

File: libsyclinterface/helper/source/dpctl_error_handlers.cpp [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+        if (!strncmp(verbose, "none", 4))
+            requested_type_level = error_level::none;
+        else if (!strncmp(verbose, "error", 5))
+            requested_type_level = error_level::error;
+        else if (!strncmp(verbose, "warning", 7))
+            requested_type_level = error_level::warning;
+    }
+
+    if (requested_type_level <= type_error) {
+        std::cerr << what << " In " << func_name << " at " << file_name << ": "
```
Comment (oleksandr-pavlyk) [inline libsyclinterface/helper/source/dpctl_error_handlers.cpp:88]:
```suggestion
        std::cerr << what << " In " << func_name << " at " << file_name << ":"
```

File: libsyclinterface/helper/include/dpctl_error_handlers.h [hunk 9]
```diff
@@ -44,3 +44,22 @@ class DPCTL_API DPCTL_AsyncErrorHandler
...
+    none = 0,
+    error = 1,
+    warning = 2
+};
+
+void error_handler(const std::exception &e,
+                   const char *file_name,
+                   const char *func_name,
+                   int line_num,
+                   int type_error = error_level::error);
```
Comment (oleksandr-pavlyk) [inline libsyclinterface/helper/include/dpctl_error_handlers.h:59]:
```suggestion
                   int error_type = error_level::error);
```

File: libsyclinterface/helper/include/dpctl_error_handlers.h [hunk 10]
```diff
@@ -44,3 +44,22 @@ class DPCTL_API DPCTL_AsyncErrorHandler
...
+
+void error_handler(const std::string &what,
+                   const char *file_name,
+                   const char *func_name,
+                   int line_num,
+                   int type_error = error_level::warning);
```
Comment (oleksandr-pavlyk) [inline libsyclinterface/helper/include/dpctl_error_handlers.h:65]:
```suggestion
                   int error_type = error_level::warning);
```

File: libsyclinterface/helper/source/dpctl_error_handlers.cpp [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+#include <cstring>
+
+void DPCTL_AsyncErrorHandler::operator()(
+    const cl::sycl::exception_list &exceptions)
+{
+    for (std::exception_ptr const &e : exceptions) {
+        try {
+            std::rethrow_exception(e);
+        } catch (cl::sycl::exception const &e) {
+            std::cerr << "Caught asynchronous SYCL exception:\n"
```
Comment (oleksandr-pavlyk) [inline libsyclinterface/helper/source/dpctl_error_handlers.cpp:36]: Perhaps we should call `error_handler` here too.

File: libsyclinterface/helper/source/dpctl_error_handlers.cpp [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+        std::cerr << e.what() << " in " << func_name << " at " << file_name
+                  << ": " << line_num << std::endl;
+    }
+}
+
+void error_handler(const std::string &what,
+                   const char *file_name,
+                   const char *func_name,
+                   int line_num,
+                   int type_error)
```
Comment (oleksandr-pavlyk) [inline libsyclinterface/helper/source/dpctl_error_handlers.cpp:73]:
```suggestion
                   int error_type)
```

File: libsyclinterface/helper/source/dpctl_error_handlers.cpp [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+    if (verbose) {
+        if (!strncmp(verbose, "none", 4))
+            requested_type_level = error_level::none;
+        else if (!strncmp(verbose, "error", 5))
+            requested_type_level = error_level::error;
+        else if (!strncmp(verbose, "warning", 7))
+            requested_type_level = error_level::warning;
+    }
+
+    if (requested_type_level <= type_error) {
```
Comment (oleksandr-pavlyk) [inline libsyclinterface/helper/source/dpctl_error_handlers.cpp:87]:
```suggestion
    if (requested_type_level <= error_type) {
```

**Review by oleksandr-pavlyk — DISMISSED** [review]
LGTM!

PR history requires cleanup though.

**Review by oleksandr-pavlyk — APPROVED** [review]
Summary of history rewriting:

First commit implements error_handler, renames error handler helper file, renames argument name for functions taking async handler callback. 

Then there are commits using `error_handler` in individual `dpct_sycl_*.cpp` files

**Pull request merged**